text/0000-per-project-config.md

```diff
@@ -0,0 +1,79 @@
+
+# Motivation
+[motivation]: #motivation
+
+We currently only support Notion configuration on a per-user basis in the `hooks.toml` file. This file allows users to change the URLs that tools are downloaded from. However, since a user could be working on both an open-source project (that uses the public registry) and an internal project (that uses a private registry), we instead want to allow the users to specify these settings on a per-project basis.
+
+Additionally, we currently already support per-project settings for the platform (node, npm, and yarn versions), specified in the `toolchain` key inside of `package.json`. Multiple users have requested an alternate method in a separate file, for various reasons (see https://github.com/notion-cli/notion/issues/282 for some discussion of one use-case). As we are already looking at adding per-project hooks to a separate file, that file should also support specifying the platform information.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+We need to document the format of the `.notionrc.json` file, which will mostly match the features available in the existing `hooks.toml`, however it will also include the additional settings for `toolchain` similar to . Additionally, we need to document and explain the precedence order for when multiple configs are specified (e.g. user-level _and_ project-level).
+
+Additionally, for consistency of configuration, we should update the user-wide config stored in `~/.notion` from `hooks.toml` to also be named `.notionrc.json`, so that it's immediately obvious that it contains the same settings as the project-level configuration. This will be a breaking change that requires users to translate their settings from the old format to the new.
+
+This feature will primarily be used for more advanced use-cases, as the default public repositories and `package.json` settings should cover the majority of open-source projects.
+
+# Details
+[details]: #details
+
+## User-level configuration file
+
+We will need to update the `HookConfig` to parse the user-level `.notionrc.json`, if it exists. We should also allow the user to specify the user platform (`node`, `npm`, and `yarn`) in that file. To that end, we can likely merge the existing `~/.notion/tools/user/platform.json` into `~/.notion/.notionrc.json` as our source of truth for the user-default selected versions. This may also allow us to simplify some of the logic around determining the currently active platform, as we can merge all of the available configs into a single result once, and then use that to execute a tool.
+
+## Configuration Precedence
+
+When configurations are specified in multiple places, we need to merge them in a specified order so that we get a single configuration to work from. The order will be defined generally as more specific configurations having precedence over more general configurations. This means:
+
+- Per-project configuration beats per-user configuration
```

Comment:
is this a deep or shallow merge

Reply:
My mental model was a deep merge (which I think shows in the example later on), I'll call that out explicitly, however.